File: ea/src/evolution_engine/evolution/BaseEvolutionStrategy.java
package evolution_engine.evolution;

import evolution_engine.fitness.FitnessCalculator;
import evolution_engine.mutators.Crossover;
import evolution_engine.mutators.Mutator;
import evolution_engine.selection.SelectionStrategy;

import java.util.List;

/**
 * Created By Itay Azaria
 * Date: 9/16/2014
 */
public abstract class BaseEvolutionStrategy<I extends Individual> implements EvolutionStrategy<I> {
    protected FitnessCalculator fitnessCalculator;
    protected SelectionStrategy selectionStrategy;
    protected Crossover<I> crossover;
    protected List<? extends Mutator> mutators;

    protected BaseEvolutionStrategy(EvolutionParameters evolutionParameters) {
        this.fitnessCalculator = evolutionParameters.getFitnessCalculator();
        this.selectionStrategy = evolutionParameters.getSelectionStrategy();
        this.crossover = evolutionParameters.getCrossover();
        this.mutators = evolutionParameters.getMutators();
    }
}

File: src/evolution_engine/evolution/BaseEvolutionStrategy.scala
package evolution_engine.evolution

import evolution_engine.fitness.FitnessCalculator
import evolution_engine.mutators.{Crossover, Mutator}
import evolution_engine.selection.SelectionStrategy

/**
 * Created By Itay Azaria
 * Date: 9/16/2014
 */
abstract class BaseEvolutionStrategy[I <: Individual](
                                                       evolutionParameters: EvolutionParameters[I]
                                                       ) extends EvolutionStrategy[I] {
  val fitnessCalculator: FitnessCalculator[I] = evolutionParameters.getFitnessCalculator
  val selectionStrategy: SelectionStrategy[I] = evolutionParameters.getSelectionStrategy
  val crossover: Crossover[I] = evolutionParameters.getCrossover
  val mutators: List[Mutator[I]] = evolutionParameters.getMutators

}